keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,858 @@
+
+	// Kubernetes 1.11 added support for specifying podSelector and namespaceSelector
+	// together in a peer.
+	NetworkPolicyVersion1_11 NetworkPolicyVersion = "1.11"
+
+	// Kubernetes 1.12 added alpha SCTP support. Plugins implementing this version are
+	// assumed to recognize `protocol: SCTP`, even if they don't implement it.
+	NetworkPolicyVersion1_12 NetworkPolicyVersion = "1.12"
+)
+
+```
+
+And add a new `Status NetworkPolicyStatus` field to `NetworkPolicy`, defined as:
+
+```
+// NetworkPolicyStatus contains information about the processing of a NetworkPolicy
+type NetworkPolicyStatus struct {
+	// conditions associated with the policy. These may include a "Supported"
+	// condition indicating whether the features used by the policy are supported,
+	// an "Enforcing" condition indicating whether the policy is actively being
+	// enforced on network traffic yet, and a "Problem" condition indicating some
+	// non-fatal problem with the policy.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []NetworkPolicyStatusCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,3,rep,name=conditions"`
+}
+
+// NetworkPolicyStatusConditionType is a valid value of NetworkPolicyStatusCondition.Type
+type NetworkPolicyStatusConditionType string
+
+const (
+	// NetworkPolicyStatusSupported - The plugin supports this NetworkPolicy.
+	NetworkPolicyStatusSupported NetworkPolicyStatusConditionType = "Supported"
+
+	// NetworkPolicyStatusProblem - The NetworkPolicy has a problem
+	NetworkPolicyStatusProblem NetworkPolicyStatusConditionType = "Problem"
+
+	// NetworkPolicyStatusEnforcing - The plugin is currently enforcing the rules
+	// specified by this NetworkPolicy
+	NetworkPolicyStatusEnforcing NetworkPolicyStatusConditionType = "Enforcing"
+)
+
+// NetworkPolicyStatusCondition contains details about the state of NetworkPolicy processing
+type NetworkPolicyStatusCondition struct {
+	Type   NetworkPolicyStatusConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=NetworkPolicyStatusConditionType"`
+	Status ConditionStatus                  `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+	// Short, machine understandable string that gives the reason for condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
+	// Human-readable message indicating details about last transition.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
+}
+```
+
+### Semantics
+
+When a user creates a NetworkPolicy, and is using a network plugin that
+implements this specification:
+
+  - If the NetworkPolicy uses API fields which are not known to the
```

Comment:
The APIserver will just discard fields it doesn't know

Reply:
Ah, so it's just `kubectl create` magic that rejects unknown fields?

OK, so in that case:

  - If the user explicitly specifies a `minVersion` that is newer than the apiserver, then we can recognize that at validation time and reject it
  - If the user doesn't explicitly specify `minVersion` and lets the apiserver default it, then the apiserver may fail to notice dropped fields and the policy may end up wrong.

So... meh, that's not useful

Reply:
> Ah, so it's just kubectl create magic that rejects unknown fields?

Yeah - it pulls the openapi doc and does some validation.